rewrite guide to pinning dependencies with niv (#782)
NixOS/nix.dev: source/guides/recipes/dependency-management.md

(dependency-management-niv)=

> # Managing remote sources with niv

Author: ```suggestion
# Automatically managing remote sources with niv
```

This is to address https://github.com/NixOS/nix.dev/pull/782/files#r1380351386, as I assume the suggestion is to emphasize *automatic*.

The Nix language can be used to describe dependencies between files managed by Nix.
Nix expressions themselves can depend on remote sources, and there are multiple ways to specify their origin, as shown in [](pinning-nixpkgs).

For more automation around handling remote sources, set up [niv](https://github.com/nmattia/niv/) in your project:

```shell-session

> $ nix-shell -p niv --run "niv init"

Author: Oops. This is the weirdest thing. Running that doesn't actually work! (To make sure, here I also specified the source for `niv`):
```shellSession
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv --run "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-23.05"
Initializing
  Creating nix/sources.nix
  Creating nix/sources.json
  Importing 'nixpkgs' ...
  Adding package nixpkgs
    FATAL: Could not read the output of 'nix-prefetch-url'. This is a bug. Please create a
    ticket:
    
      https://github.com/nmattia/niv/issues/new
    
    Thanks! I'll buy you a beer.
    command: 
    nix-prefetch-url--unpack https://github.com/nixos/nixpkgs/archive/34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz --name 34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz
    stdout: 
    
    stderr: 
    error: failed to extract archive (Write failed)
    
    FATAL: One or more packages failed to update:
    nixpkgs: ExitFailure 1
  ERROR: ExitFailure 1
ERROR: ExitFailure 1
```

The command sample that is output *looks* broken, but it's just [a little representation bug here](https://github.com/nmattia/niv/blob/723f0eeb969a730db3c30f977c2b66b9dce9fe4a/src/Niv/GitHub/Cmd.hs#L187):

```shellSession
$ nix-prefetch-url --unpack https://github.com/nixos/nixpkgs/archive/34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz --name 34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz
```
Running this in the shell from which we originally called `nix-shell` will in fact succeed. But that's a different environment. If we try the pure shell, and drop into it [during the `--command`](https://nixos.org/manual/nix/stable/command-ref/nix-shell#options), to attempt to run that command, it will fail (don't forget to clean the `nix` directory left over from the previous call):

 ```shellSession
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --command "niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-23.05; return"
# ... same error as above
[nix-shell]$ nix-prefetch-url --unpack https://github.com/nixos/nixpkgs/archive/34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz --name 34bdaaf1f0b7fb6d9091472edc968ff10a8c2857.tar.gz
error: failed to extract archive (Write failed)
```

Without looking into the C++ I'd guess that something about the environment in `nix-shell` is different that prevents `nixpkgs-prefetch-url` to write to the store. Even weirder is that this is only true for `NixOS/nixpkgs`, but not for a smaller repo such as used in https://github.com/NixOS/nix/issues/5614.

Proof. For a fair comparison, init with empty sources and then add `nixpkgs` (fails), then start over and add `haskell.nix`:

```shellSession
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --run "niv init --no-nixpkgs; niv add nixos/nixpkgs -b nixos-23.05"
# ... same error
```

```shellSession
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --run "niv init --no-nixpkgs; niv add nixos/nixpkgs -b nixos-23.05"
# ... same error
$ rm -rf nix
$ nix-shell -I nixpkgs=channel:nixos-23.05 -p niv nix cacert --pure --run "niv init --no-nixpkgs; niv add input-output-hk/haskell.nix -r aaf4e230c69507fb3a1c6fe4c1f6603d659d5543"
# ...
Done: Adding package haskell.nix
```

One must assume there is something in `nix-prefetch-url` that will behave differently with a large download in a `nix-shell` environment.

> $ nix-shell -p niv --run "niv init"

Author: Oh wow, I indeed ran into `/tmp` being full!

Thanks to these two helpful comments this was not entirely impossible to debug:
- https://github.com/DavHau/mach-nix/issues/458#issuecomment-1248583956
- https://github.com/NixOS/nix/issues/5857#issuecomment-1509862011

The funniest part is the reason how it even got there: When we tested the kernel build instructions with @StefanSchroeder earlier, the kernel *source code* filled up my `tmpfs` with 1.5GiB. Now that's a massive codebase.

> $ nix-shell -p niv --run "niv init"

Reviewer: Damn, nice find! I guess in the end it's unrelated and we shouldn't worry about it :)

```

This command will generate `nix/sources.json` in the current directory, which is a lock file for dependencies.
It will also create `nix/sources.nix`, which exposes those dependencies as an attribute set.

:::{note}
By default, `niv init` will add the latest revision of `nixpkgs-unstable` as a source.
If you need the latest revision of a specific branch:

```shell-session
niv init --nixpkgs nixos/nixpkgs --nixpkgs-branch nixos-23.05
```
:::

> Import the generated `nix/sources.nix` for the top-level argument in the top-level `default.nix` and use it to refer to the Nixpkgs source directory:

Author: ```suggestion
Import the generated `nix/sources.nix` for the top-level argument in `default.nix` and use it to refer to the Nixpkgs source directory:
```

```nix
{ sources ? import ./nix/sources.nix }:
let
  pkgs = import sources.nixpkgs {};
  build = pkgs.hello;
in {
  inherit build;
}
```

> `nix-build` will call the top-level function with the default argument.

Reviewer: Technically `nix-build` just calls it with `{}` (or the additional `--arg`/`--argstr` values).

Author: ```suggestion
`nix-build` will call the top-level function with the empty attribute set `{}`, or with the attributes passed via [`--arg`](https://nixos.org/manual/nix/stable/command-ref/nix-build#opt-arg) or [`--argstr`](https://nixos.org/manual/nix/stable/command-ref/nix-build#opt-argstr).
```

This pattern allows [overriding remote sources](overriding-sources-niv) programmatically.

Add niv to the development environment for your project to have it readily available:

```diff
 { sources ? import ./nix/sources.nix }:
 let
   pkgs = import sources.nixpkgs {};
   build = pkgs.hello;
 in {
   inherit build;

> +  shell = pkgs.mkShell {

Reviewer: Let's also have a simple `shell.nix` that works for this `default.nix`

+    inputsFrom = [ build ];
+    packages = with pkgs; [
+      niv
+    ];
+  };
 }
```

> See [](./sharing-dependencies) for details.

Author: ````suggestion
Also add a `shell.nix` to enter that environment more conveniently:

```nix
(import ./. {}).shell
```

See [](./sharing-dependencies) for details, and note that here you have to pass an empty attribute set to the imported expression, since `default.nix` now contains a function.
````
